src/hotspot/share/logging/logTagSet.cpp

```diff
@@ -73,15 +73,19 @@ bool LogTagSet::has_output(const LogOutput* output) {
 }
 
 void LogTagSet::log(LogLevelType level, const char* msg) {
+  LogOutputList::Iterator it = _output_list.iterator(level);
   LogDecorations decorations(level, *this, _decorators);
-  for (LogOutputList::Iterator it = _output_list.iterator(level); it != _output_list.end(); it++) {
+
+  for (; it != _output_list.end(); it++) {
```

Comment:
The reordering here needs a comment, else someone will be tempted to restore it to the original form.

Reply:
I understand there is a race. I'm asking you to add a comment explaining there is a race so that someone is not tempted to reorder these statements again. That said it seems to me that we also require a `loadload` barrier between these (which would pair with the storestore barrier in `LogConfiguration::configure_output`).

Reply:
I see. I would like to convince reviewers first. I will add some comments here. 

I don't think loadload barrier is necessary.  volatile in c/c++ is a little bit different from Java. we can't fully apply JSR-133 here. Yes, we do need to make sure Atomic::add() happens before load of decorators. This is done by the implied memory barrier of `Atomic::add`.

Reply:
I have no idea why you mention volatile and jsr-133 here.

If there is an Atomic::add inside there then okay the necessary barrier may be in place. But it is not at all apparent there is an Atomic operation inside there, so given the order is important the comment needs to explain why an explicit barrier is not needed.

Reply:
The following code uses `Atomic::add`. Actually, there's a hidden 3rd argument. it's a default parameter and I highlight it.  It is the conservative memory order emits a full memory barrier. I double-check the implementations on  x86/aarch64 and powerpc.

```
jint LogOutputList::increase_readers() {
  jint result = Atomic::add(&_active_readers, 1, /*memory_order_conservative*/);
  assert(_active_readers > 0, "Ensure we have consistent state");
  return result;
}
```

Reply:
Thanks for adding the comment.